File: src/Node/parser.js
const instr =           require('instructions.js');
const instruction_idx = instr.instruction_idx;

function parseNumber(src) {
	if (src.match(".")) return parseFloat(src);
	return parseInt(src);
}

const isNumber = function(src) {return parseNumber(src) != NaN;}
const isLetter = function(src) {return src.match(/[a-z]/) != null;}

function getProtos(src) {
	src = src + "\n";
	const len =  #src;
	var v =      0;
	const add =  function(x) {v = v + (x || 1);}
	const next = function() {add(); return src.slice(v - 1, v);}
	function getLines() {
		var line = "";
		var lines = [];
		while(true) {
			if (v > len) break;
			const sub = next();
			if (sub === "\n") {
				if (line != "") lines[lines.length] = line;
			} else {
				line = line + sub;
			}
		}
		return lines;
	}
	const lines = getLines();
	var found =   false;
	var proto =   {name: "", instructions: []};
	var protos =  [];
	for (var idx = 0; idx < lines.length; idx++) {
		const line = lines[idx];
		const last = line.slice(-1);
		const rest = line.slice(0, -2):
		if (last === ":") {
			if (!found) {
				proto.name = rest;
				found =      true;
			} else {
				protos[protos.length] = proto;
				proto =                 {name: "", instructions: []};
				proto.name =            rest;
			}
		}
	}
	return protos;
}

function parse(src) {
	var protos = getProtos(src);
	for (var idx = 0; idx < protos.length; idx++) {
		var proto =        protos[idx];
		var name =         proto.name;
		var instructions = proto.instructions;
		for (var _idx = 0; _idx < instructions.length; _idx++) {
			var instruction = instructions[_idx];
			var _new =        [];
			var len =         instruction.length;
			var pos =         0;
			var opFound =     false;
			while(true) {
				pos++;
				if (pos > len || _new > 4) break;
				const sub = instruction.slice(pos - 1, pos);
				if (isLetter(sub)) {
					pos--;
					if (!opFound) {
						var op = "";
						while(true) {
							pos++;
							if (pos > len) break;
							const _sub = instruction.slice(pos - 1, pos);
							if (isLetter(_sub)) {
								op = op + sub;
							} else {
								break;
							}
						}
						opFound = true;
						_new[1] = instruction_idx[op];
					} else {
						var arg = "";
						while(true) {
							pos++;
							if (pos > len) break;
							const _sub = instruction.slice(pos - 1, pos);
							if (isLetter(_sub) || isNumber(_sub) || _sub === "_") {
								arg = arg + sub;
							} else {
								break;
							}
						}
						_new[_new.length] = arg;
					}
				} else if (isNumber(sub)) {
					pos--;
					if (instruction.slice(pos - 1, pos) === "-") arg = arg .. "-";
					var arg = "";
					while(true) {
						pos++;
						if (pos > len) break;
						const _sub = instruction.slice(pos - 1, pos);
						if (isNumber(_sub) || _sub === ".") {
							arg = arg + sub;
						} else {
							break;
						}
					}
					_new[_new.length] = parseNumber(arg);
				}
			}
			instructions[_idx] = _new;
		}
	}
	return protos;
}

module.exports = {parse: parse};

File: src/Parser.lua
-- 8ch_32bit

local string = string;
local table = table;

local InstructionIdx = require('Instructions').InstructionIdx;

local function Parse(Source)
	if not Source then
		return;
	end;
	
	Source = Source .. "\n";

	local Found        = false;
	local Proto        = {};
	local Instructions = {};
	local Protos       = {};
	local Lines        = {};
	
	local Length   = #Source;
	local Position = 0;
	
	local CurrentLine = "";

	::continue::
		
	while Position <= Length do
		Position = Position + 1;
		local Substring = string.sub(Source, Position, Position);
			
		if CurrentLine ~= "" then
			if Substring == "\n" then
				CurrentLine = CurrentLine .. Substring;

				goto continue;
			end;
		end;
		
		table.insert(Lines, CurrentLine);
				
		CurrentLine = "";
	end;

	local LineCount = #Lines;
	
	for Idx = 1, LineCount, 1 do
		local Line = Lines[Idx];
		
		local Last = string.sub(Line, -1);
		local Rest = string.sub(Line, 1, -2);

		if Last == ";" or Last == "\n" then
			table.insert(Instructions, string.lower(Rest));
		end;
		
		if Last == ":" then
			if not Found then
				Proto.Name = Rest;
				Found      = true;
			else
				Proto.Instructions = Instructions;
				table.insert(Protos, Proto);
				
				Proto        = { Name = Rest };
				Instructions = {};
			end;
		end;
		
		if Idx == LineCount then
			Proto.Instructions = Instructions;
			table.insert(Protos, Proto);
		end;
	end;
	
	local ProtoCount = #Protos;
	
	for Idx = 1, ProtoCount, 1 do
		local Proto        = Protos[idx];
		local Name         = Proto.Name;
		local Instructions = Proto.Instructions;
		local InstrCount   = #Instructions
		
		for _idx = 1, #instructions do
			local Instruction = Instructions[_idx];
			local Length      = #Instruction;
			local Position    = 0;
			local New         = {};
			
			local OpcodeFound = false;
			
			while Position <= Length do
				Position = Position + 1;
				
				local Substring = string.sub(Instruction, Position, Position);
				
				if string.find(Substring, "[a-z]") then
					Position = Position - 1;
					
					if OpcodeFound then
						local Arg = "";
						
						while Position <= Length do
							Position = Position + 1;
							local Substring2 = string.sub(Instruction, Position, posPosition
							
							if string.find(Substring2, "[a-z]") or tonumber(Substring2) or Substring2 == "_" then
								Arg = Arg .. Substring2;
							else
								break;
							end;
						end;
						
						table.insert(New, Arg);
					else
						local Opcode = "";
						
						while Position <= Length do
							Position = Position + 1;
							local Substring2 = string.sub(Instruction, Position, Position);
							
							if isLetter(Substring2) then
								Opcode = Opcode .. _sub;
							else
								break;
							end;
						end;
						
						OpcodeFound = true;
						New[1] = InstructionIdx[Opcode];
					end;
				elseif tonumber(Substring) then
					Position = Position - 1;
					
					local Arg = "";
					
					if string.sub(Instruction, Position, Position) == "-" then
						Arg = Arg .. "-";
					end
					
					while Position <= Length do
						Position = Position + 1;
						local Substring2 = string.sub(Instruction, Position, Position);
						
						if tonumber(Substring2) or Substring2 == "." then
							Arg = Arg .. Substring2;
						else
							break;
						end;
					end;
					
					table.insert(New, tonumber(Arg) or Arg);
				end;

				if #New > 4 then
					break;
				end;
			end
			
			instructions[_idx] = new;
		end
	end
	
	return protos;
end;

return {parse = parse};
